src/cryptsetup/cryptsetup.c

```diff
@@ -2440,33 +2459,40 @@ static int run(int argc, char *argv[]) {
                 }
 #endif
 
-                bool use_cached_passphrase = true;
+                bool use_cached_passphrase = true, try_discover_key = !key_file;
+                _cleanup_free_ char *discovered_key_fn = strjoin(volume, ".key");
```

Comment:
oom check missing

Reply:
Since we moved this out of any loop I brought back the `strjoina()` call.